exonum/src/events/noise/wrappers/sodium_wrapper/handshake.rs

```diff
@@ -140,25 +189,38 @@ impl NoiseHandshake {
 }
 
 impl Handshake for NoiseHandshake {
-    fn listen<S>(self, stream: S) -> HandshakeResult<S>
+    type Result = Vec<u8>;
+
+    fn listen<S>(self, stream: S) -> HandshakeResult<S, Self::Result>
     where
         S: AsyncRead + AsyncWrite + 'static,
     {
         let framed = self.read_handshake_msg(stream)
-            .and_then(|(stream, handshake)| handshake.write_handshake_msg(stream))
+            .and_then(|(stream, handshake, _)| handshake.write_handshake_msg(stream, &[]))
             .and_then(|(stream, handshake)| handshake.read_handshake_msg(stream))
-            .and_then(|(stream, handshake)| handshake.finalize(stream));
+            .and_then(|(stream, handshake, peers_exchange)| {
+                handshake.finalize(stream, peers_exchange)
+            });
         Box::new(framed)
     }
 
-    fn send<S>(self, stream: S) -> HandshakeResult<S>
+    fn send<S>(self, stream: S) -> HandshakeResult<S, Self::Result>
     where
         S: AsyncRead + AsyncWrite + 'static,
     {
-        let framed = self.write_handshake_msg(stream)
-            .and_then(|(stream, handshake)| handshake.read_handshake_msg(stream))
-            .and_then(|(stream, handshake)| handshake.write_handshake_msg(stream))
-            .and_then(|(stream, handshake)| handshake.finalize(stream));
-        Box::new(framed)
+        match self.peers_exchange.clone() {
+            Some(peers_exchange) => {
+                let framed = self.write_handshake_msg(stream, &[])
+                    .and_then(|(stream, handshake)| handshake.read_handshake_msg(stream))
+                    .and_then(move |(stream, handshake, _)| {
+                        handshake.write_handshake_msg(stream, &peers_exchange.into_bytes())
+                    })
```

Comment:
This ugly formatting and `move` closure trigger me with inconsistency, but I don't see how this can be improved :(

Reply:
I don't like it either(